docs/bespoke.md

```diff
@@ -0,0 +1,121 @@
+
+
+### Per-environment settings
+
+The site local settings file includes the alias definition, binding it to the
+actual database credentials, allowing for per-environment configuration:
+
+    <?php
+    // settings.local.php
+    $settings['mongodb'] = [
+      'clients' => [
+        // Client alias => constructor parameters.
+        'default' => [
+          'uri' => 'mongodb://localhost:27017',
+          'uriOptions' => [],
+          'driverOptions' => [],
+        ],
+      ],
+      'databases' => [
+        // Collection alias => [ client_alias, collection_name ]
+        'default' => ['default', 'drupal'],
+        'logger' => ['default', 'logger'],
+        'foodb' => ['default', 'foo-database'],
+      ],
+    ];
+
+With such a configuration, the `foodb` alias is available to all MongoDB-using
+modules in the site, possibly pointing to different databases depending on the
+environment (development, staging, production...).
+
+
+### Service-based module adapter
+
+The `foo.services.yml` service file for the bespoke `foo.module` can then
+reference `foodb` to access the database with a constant alias, regardless
+of the environment:
+
+    // modules/custom/Foo/foo.services.yml
+    services:
+      foo.storage:
+        class: 'MongoDB\Database'
+        factory: ['@mongodb.database_factory', 'get']
+        arguments: ['foodb']
+
+      foo.bar:
+        class: 'Drupal\foo\Bar'
+        arguments: ['@foo.storage', '$logger.channel.foo']
+
+This allows services in the module to access the database in both function code
+for Drupal hooks, and OO code for component-level logic without having to be
+environment-aware.
+
+If the `mongodb_watchdog` module is enabled, the logger passed to the
+application will be a standard PSR-3 logger writing to MongoDB without the
+application having to know anything about it, but still providing a standard
+Drupal UI to examine the application logs.
+
+
+### Component logic
+
+Finally the component application logic can use the services without receiving
+any Drupal-specific dependency. In this example, we can simply assume the
+service code is located within the module itself, for simplicity:
+
+    <?php
+    // modules/custom/Foo/src/Bar.php
+    use MongoDb\Database;
+    use Psr\Log\LoggerInterface;
+
+    public function __construct(Database $database, LoggerInterface $logger) {
+      $this->database = $database;
+      $this->logger = $logger;
+    }
+
+    public function baz() {
+      // Perform some business logic using $this->database.
+      // Log it using $this->logger.
+    }
+
+Having the code only receive standard services (like a PSR-3 logger) or
+[PHP library] classes allows it to be written as an agnostic component that can
+be brought in using Composer and shared with non-Drupal code. This is often
+useful in bespoke projects, which tend to combine Drupal with other parts of the
+application written in Laravel or Symfony standard edition, since the code has
```

Comment:
Symfony Standard Edition is now retired https://symfony.com/blog/the-end-of-the-symfony-standard-edition so we should reference something else. Maybe the Symfony Flex symfony/website-skeleton ?